domain/credential/state/credential.go

```diff
@@ -0,0 +1,489 @@
+		return nil, errors.Trace(err)
+	}
+
+	var creds []cloud.Credential
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		var err error
+		creds, err = st.loadCloudCredentials(ctx, tx, "", cloudName, owner)
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	result := make(map[string]cloud.Credential)
+	for _, cred := range creds {
+		result[cred.Label] = cred
+	}
+	return result, nil
+}
+
+// CloudCredential returns the cloud credential for the given tag.
+func (st *State) CloudCredential(ctx context.Context, name, cloudName, owner string) (cloud.Credential, error) {
+	db, err := st.DB()
+	if err != nil {
+		return cloud.Credential{}, errors.Trace(err)
+	}
+
+	var creds []cloud.Credential
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		var err error
+		creds, err = st.loadCloudCredentials(ctx, tx, name, cloudName, owner)
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return cloud.Credential{}, errors.Trace(err)
+	}
+	if len(creds) == 0 {
+		return cloud.Credential{}, fmt.Errorf("credential %q for cloud %q owned by %q %w", name, cloudName, owner, errors.NotFound)
+	}
+	if len(creds) > 1 {
+		return cloud.Credential{}, errors.Errorf("expected 1 credential, got %d", len(creds))
+	}
+	return creds[0], errors.Trace(err)
+}
+
+type credentialInfo struct {
+	id            string
+	name          string
+	revoked       bool
+	invalid       bool
+	invalidReason string
+	authType      string
+	cloudName     string
+	owner         string
+}
+
+type credentialAttrs map[string]string
+
+func (st *State) loadCloudCredentials(ctx context.Context, tx *sql.Tx, name, cloudName, owner string) ([]cloud.Credential, error) {
+	// First load the basic credential info.
+	credQuery := `
+SELECT cloud_credential.uuid, cloud_credential.name,
+       cloud_credential.revoked, cloud_credential.invalid, 
+       cloud_credential.invalid_reason, 
+       cloud_credential.owner_uuid AS owner,
+       auth_type.type AS auth_type, cloud.name AS cloud_name,
+       cloud_credential_attributes.key, cloud_credential_attributes.value
+FROM   cloud_credential
+       INNER JOIN auth_type 
+             ON cloud_credential.auth_type_id = auth_type.id
+       INNER JOIN cloud 
+             ON cloud_credential.cloud_uuid = cloud.uuid
+       LEFT JOIN cloud_credential_attributes
+            ON cloud_credential_attributes.cloud_credential_uuid = cloud_credential.uuid
+`
+
+	var (
+		args        []any
+		filterTerms []string
+	)
+	if name != "" {
+		filterTerms = append(filterTerms, "cloud_credential.name = ?")
+		args = append(args, name)
+	}
+	if cloudName != "" {
+		filterTerms = append(filterTerms, "cloud.name = ?")
+		args = append(args, cloudName)
+	}
+	if owner != "" {
+		filterTerms = append(filterTerms, "cloud_credential.owner_uuid = ?")
+		args = append(args, owner)
+	}
+	if len(filterTerms) > 0 {
+		credQuery = credQuery + "\nWHERE " + strings.Join(filterTerms, " AND ")
+	}
```

Comment:
This feels like it could be extracted into a nice pattern.

Reply:
Make a helper function.